driver-async/src/main/com/mongodb/async/client/ClientSessionImpl.java

```diff
@@ -74,18 +71,28 @@ public void startTransaction(final TransactionOptions transactionOptions) {
 
     @Override
     public void commitTransaction(final SingleResultCallback<Void> callback) {
-        if (!inTransaction) {
+        if (!canCommitOrAbort()) {
             throw new IllegalStateException("There is no transaction started");
         }
-        endTransaction(new CommitTransactionOperation(transactionOptions.getWriteConcern()), callback);
+        endTransaction(new TransactionOperationCreator() {
+            @Override
```

Comment:
Instead of passing in a *function* to create the operation why not just pass in the operation?

You can still centralize in a helper method the creation of the operation.

Reply:
Sorry, should have explained that.  If you pass in the operation, you get a NullPointerException on transactionOptions when autoStart is true and the transaction hasn't actually been started yet:

     try (var s = client.startSession()) {
         // do nothing
         s.commitTransaction();
    }

I could refactor but this way allows a bit more code-sharing between commit and abort.

Reply:
So `transactionOptions` are mutable and at the time of `commitTransaction()` in the example above it hasn't been set? 

The code doesn't read that way and I think would be clearer if the `writeConcern` was explicitly passed into the `create` method rather than relying on mutable global state ensuring its there when its called.